### PR TITLE
[ENG-342] Block location watcher for files that are being actively modified

### PR DESCRIPTION
**Pull request by fogodev**
Introducing methods to temporarily stops and reinit watching over a location

### CODE REVIEW

**Comment by Brendonovich** [comment]
Does this PR address the entirety of the linked issue? Surely while encrypting a file we don't have to stop watching the location _entirely_, just ignore events for the files being processed.

**Comment by fogodev** [comment]
> Does this PR address the entirety of the linked issue? Surely while encrypting a file we don't have to stop watching the location _entirely_, just ignore events for the files being processed.

Just introduced this part of ignoring events based on a path, used a guard to temporarily ignore those events.

File: core/src/location/manager/helpers.rs [hunk 1]
```diff
@@ -163,3 +166,137 @@ pub(super) fn subtract_location_path(
...
+						locations_watched,
+						locations_unwatched,
+					);
+					forced_unwatch.remove(&key);
+
+					Ok(())
+				} else {
+					Err(LocationManagerError::LocationMissingLocalPath(location_id))
+				}
+			} else {
```
Comment (brxken128) [inline core/src/location/manager/helpers.rs:279]: a tad messy (along with the other one), but if it's too much hassle to change it'll be fine

File: core/src/location/manager/helpers.rs [hunk 2]
```diff
@@ -163,3 +166,137 @@ pub(super) fn subtract_location_path(
...
+				Err(LocationManagerError::FailedToStopOrReinitWatcher {
+					reason: String::from("failed to fetch location from db"),
+				})
+			}
+		} else {
+			Ok(())
+		},
+	); // ignore errors, we handle errors on receiver
```
Comment (brxken128) [inline core/src/location/manager/helpers.rs:287]: a tad messy, but if it's too much hassle to change then not an issue (same with the other similar pattern)

**Review by brxken128 — APPROVED** [review]
looks great, haven't tested but provided you have i don't see anything that should really break (and it fully addresses the issue, which is a major plus in my eyes haha)

File: core/src/location/manager/helpers.rs (hunk 2 above)
Comment (fogodev) [reply core/src/location/manager/helpers.rs]: Changed to a more functional style

**Review by Brendonovich — APPROVED** [review]
I wonder if we could use something like Actix or build our own actor abstraction to simplify the actors & message channels, it's a bit unwieldy atm 😅

**Comment by jamiepine** [comment]
> Also, could the location watching be moved to its own crate?

Potentially, but I think it's good to merge for now.

**Review by jamiepine — DISMISSED** [review]
Great work!

File: core/src/location/manager/mod.rs [hunk 3]
```diff
@@ -115,10 +126,48 @@ impl LocationManager {
...
+			Ok(())
+		}
+	}
+
 	#[cfg(feature = "location-watcher")]
 	async fn run_locations_checker(
 		mut add_locations_rx: mpsc::Receiver<ManagerMessage>,
 		mut remove_locations_rx: mpsc::Receiver<ManagerMessage>,
+		mut stop_watcher_rx: mpsc::Receiver<ManagerMessage>,
+		mut reinit_watcher_rx: mpsc::Receiver<ManagerMessage>,
```
[the remark this answers is not in the document]
Comment (fogodev) [reply core/src/location/manager/mod.rs]: I agree with you, just did some refactoring to introduce nicer abstractions and unifying those channels in a more extensible way, so it's easier to introduce more features in the future.

**Review by Brendonovich — APPROVED** [review]

**Pull request merged**